Draft about a proposal for nested multiline imports

benjaminabel opened the pull request:
This is only a draft to see if you'd like to add this feature, tests haven't been changed. What do you think @miki725 ?

TODO

- [ ] add `--formatter` option in the script
- [ ] refactor `formatted()` tests to `test_formatters.py`
- [ ] add new tests for new formatter
- [ ] add some docs with examples of difference in formatters
- [ ] update `README`

miki725 commented on tests/test_data/output.txt:
```diff
@@ -8,28 +8,30 @@ import something  # with comment
-    foo,
-    rainbows,
-)
+                                                 CONSTANT,
```
not sure if this would be expected output. I would expect something like:

```python
from package.subpackage.module.submodule import (CONSTANT,
                                                 Klass,
                                                 bar,
                                                 foo,
                                                 rainbows)
```

miki725 commented:
Thank you for the proposal. I am not personal fan of this style but I know other people who are. Anyway, this will require a generic solution where even more styles will be supported. I have this feature on the roadmap however not sure when will be able to do it - #12

benjaminabel commented:
You're right we should treat first import differently and also add more introspection instead of reformating after if lines exceed 80 lines.
How would you like the styles be chosen by user: command-line? Importanize.rc file?

miki725 commented:
Both?
On Feb 19, 2015 7:40 AM, "Benjamin ABEL" <notifications@github.com> wrote:

> You're right we should treat first import differently and also add more
> introspection instead of reformating after if lines exceed 80 lines.
> How would you like the styles be chosen by user: command-line?
> Importanize.rc file?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/miki725/importanize/pull/23#issuecomment-75045514>.
>

miki725 commented:
I am not sure when Ill be able to work on it so if you want you can start. Here were my thoughts for the implementation.

1 Gut the [`formatted()`](https://github.com/benjaminabel/importanize/blob/nested-multiline-imports/importanize/statements.py#L138-L203) function.
2 Instead add a module `formatters.py` which will implement different formatters as classes:
```python
class VerticalHangingFormatter(object):
    def __init__(self, statement):
        ...
    def format(self):
        # move existing formatted implementation here
```
3 Add a parameter to statement/group/etc to accept a formatter class to use.
4 Replace `formatted()` to use the formatter:

```python
def formatted(self):
    return self.formatter.format()
```

benjaminabel commented:
Thanks for these elements. I'll try to work on this next week.

miki725 commented on importanize/formatters.py:
```diff
@@ -0,0 +1,139 @@
+        self.statement = statement
+
+
+class IndentWithTabsFormatter(Formatter):
```
can you add examples for each formatter in the docstring?

benjaminabel replied:
Actually my changes are not finalised, I just pushed to check if tests were OK in Travis. **Thanks for the review**. I'll add docstrings when my formatters will be cleaner.

miki725 replied:
cool. thx

miki725 commented on tox.ini:
```diff
@@ -6,10 +6,11 @@ setenv =
 
 [flake8]
 exclude = tests/test_data/*
+ignore = E731,W503
```
why add these?

benjaminabel replied:
This is necessary because latest pep8 version added some warnings, E731 cares about lamda expressions. This need to change existing code.

miki725 commented on tox.ini:
```diff
@@ -6,10 +6,11 @@ setenv =
     make check
-deps =
-    -rrequirements-dev.txt
+deps = -rrequirements.txt
```
does this bare any difference in tox?

benjaminabel replied:
No, it doesn't, it isn't necessary. I was just surprised by the syntax without space.

miki725 commented:
@benjaminabel :+1: overall looks good. just some tests will need to be moved/created.

miki725 commented:
just added some TODOs in the PR description. Please note that you do not have to work on all TODOs. I just want it listed to keep track of them. When Ill have a chance Ill create create another branch where Ill pull your changes plus Ill make my changes there as well. Then you can open PR against branch.

benjaminabel commented:
@miki725 I added a `test_formatter` file in which I've modified the `_test function` to accept a formatter. For now, the 2 formatters are just a copy of the initial one. I stay in this branch for now, or you prefer to use the new formatter one?

benjaminabel commented:
I created a new branch [`formatters`](https://github.com/benjaminabel/importanize/tree/formatters) for this PR, and added a formatter and tests for it.
